### PR TITLE
[Micro Gateway] Components required to configure Micro Gateway with WSO2 API Cloud

### PR DESCRIPTION
**Pull request by Dilhasha**
### Proposed changes in this pull request

This PR adds two new components and modifies an existing component to Micro Gateway implementation.
1. **org.wso2.carbon.apimgt.micro.gateway.configurator** : Configures Micro Gateway to work with WSO2 API Cloud
2. **org.wso2.carbon.apimgt.micro.gateway.status.checker** : schedules a ping request to a specified endpoint from the Micro Gateway. This can be configured to ping any endpoint with a unique identifier. 
3. **org.wso2.carbon.apimgt.micro.gateway.common** : modified to add some globally used constants

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/java/org/wso2/carbon/apimgt/micro/gateway/configurator/TestUtil.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        this.setCarbonHome();
+    }
+
+    private void setCarbonHome() throws Exception {
+        if (StringUtils.isEmpty(System.getProperty("carbon.home"))) {
+            String filePath =
+                    "src" + File.separator + "test" + File.separator + "resources" + File.separator + "carbon-home";
+            File file = new File(filePath);
+            if (file.exists()) {
+                System.setProperty("carbon.home", file.getAbsolutePath());
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/java/org/wso2/carbon/apimgt/micro/gateway/configurator/TestUtil.java]: Let's move carbon.home to a constant.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/carbon.xml [hunk 2]
```diff
@@ -0,0 +1,684 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<!--
+ Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/carbon.xml]: Please update the licence year.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml [hunk 3]
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="UTF-8"?><!--
+  ~ Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+  ~
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml]: Update here as well.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/registry.xml [hunk 4]
```diff
@@ -0,0 +1,370 @@
+<?xml version="1.0" encoding="UTF-8"?><!--
+  ~ Copyright 2005-2011 WSO2, Inc. (http://wso2.com)
+  ~
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/registry.xml]: Update the year.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.status.checker/pom.xml [hunk 5]
```diff
@@ -0,0 +1,158 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.status.checker/pom.xml]: Shall we add the licence header to this file?

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/carbon.xml (hunk 2 above)
Comment (Dilhasha) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/carbon.xml]: These are not files that I have implemented, this is maintained as a resource for testing purposes. So, updating the license year is not required IMO.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml (hunk 3 above)
Comment (Dilhasha) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml]: These are not files that I have implemented, this is maintained as a resource for testing purposes. So, updating the license year is not required IMO.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/registry.xml (hunk 4 above)
Comment (Dilhasha) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/registry.xml]: These are not files that I have implemented, this is maintained as a resource for testing purposes. So, updating the license year is not required IMO.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/carbon.xml (hunk 2 above)
Comment (npamudika) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/carbon.xml]: Although these are reused, as we have added them into the resources under this component, I think it is good if we can update the licence year.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/main/java/org/wso2/carbon/apimgt/micro/gateway/configurator/Configurator.java [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+     * Generate authentication header
+     *
+     * @param args
+     * @return
+     * @throws IOException
+     */
+    protected static String createAuthHeader(String[] args) throws IOException {
+        //Order of args - email, tenantDomain, password
+        String username = args[0] + OnPremiseGatewayConstants.USERNAME_SEPARATOR + args[1];
+        String password = args[2];
```
Comment (sumedhassk) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/main/java/org/wso2/carbon/apimgt/micro/gateway/configurator/Configurator.java]: Better to use a char[] to hold the password and clean it after using it.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.status.checker/pom.xml [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                        <Import-Package>
+                            javax.net.ssl.*,
+                            org.wso2.carbon.core; version="${carbon.kernel.package.import.version.range}",
+                            org.apache.commons.logging.*; version="${commons.logging.imp.pkg.version}",
+                            org.apache.commons.io.*,
+                            org.apache.commons.lang.*,
+                            org.osgi.service.component.*,
+                            org.osgi.framework.*,
+                            com.fasterxml.jackson.*,
+                            org.apache.commons.codec.binary.*,
```
Comment (sumedhassk) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.status.checker/pom.xml]: Better to have correct version ranges for all the imports

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.status.checker/src/main/java/org/wso2/carbon/apimgt/micro/gateway/status/checker/StatusChecker.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    /**
+     * Method to override run
+     */
+    @Override public void run() {
+        //Get credentials
+        APIManagerConfiguration config = ServiceReferenceHolder.getInstance().getAPIManagerConfigurationService()
+                                                               .getAPIManagerConfiguration();
+        String username = config.getFirstProperty(APIConstants.API_KEY_VALIDATOR_USERNAME);
+        String password = config.getFirstProperty(APIConstants.API_KEY_VALIDATOR_PASSWORD);
```
Comment (sumedhassk) [inline components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.status.checker/src/main/java/org/wso2/carbon/apimgt/micro/gateway/status/checker/StatusChecker.java]: Better to handle password with char[] and then clean it after using.

File: components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml [hunk 9]
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="UTF-8"?><!--
+  ~ Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Dilhasha) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml]: @mushthaq33 this is the license header available in APIM 2.1.0. Does this get updated in 2.2.0 release?

[the remark this answers is not in the document]
Comment (mushthaq33) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml]: @Dilhasha  But when you add this as a new file to be used by test cases, we need to update the license header year.

[the remark this answers is not in the document]
Comment (Dilhasha) [reply components/apimgt/org.wso2.carbon.apimgt.micro.gateway/org.wso2.carbon.apimgt.micro.gateway.configurator/src/test/resources/carbon-home/repository/conf/multitenancy/stratos.xml]: Ok. My understanding was different. I'll update the license header year every where.

**Pull request merged**